chia/wallet/wallet.py

```diff
@@ -394,6 +399,19 @@ async def _generate_unsigned_transaction(
             # Only one coin creates outputs
             if origin_id in (None, coin.name()):
                 origin_id = coin.name()
+                public_key = await self.hack_populate_secret_key_for_puzzle_hash(coin.puzzle_hash)
+                inner_puzzle = puzzle_for_pk(public_key)
+                decorated_target_puzzle_hash = decorator_manager.decorate_target_puzzle_hash(
+                    inner_puzzle, newpuzzlehash
+                )
+                target_primary: List[Payment] = []
+                if memos is None:
+                    memos = []
+                memos = decorator_manager.decorate_memos(inner_puzzle, newpuzzlehash, memos)
+                if (primaries_input is None and amount > 0) or primaries_input is not None:
+                    primaries.append(Payment(decorated_target_puzzle_hash, amount, memos))
+                    target_primary.append(Payment(newpuzzlehash, amount, memos))
+
```

Comment:
```suggestion
                if len(primaries) > 0 and primaries_input is None:
                    target_primary.append(primaries[0])
                    decorated_target_puzzle_hash = decorator_manager.decorate_target_puzzle_hash(
                        inner_puzzle, primaries[0].puzzle_hash
                    )
                    decorated_memos = decorator_manager.decorate_memos(
                        inner_puzzle, primaries[0].puzzle_hash, primaries[0].memos
                    )
                    decorated_payment = Payment(decorated_target_puzzle_hash, amount, decorated_memos)
                    if primaries[0] != decorated_payment:
                        self.log.debug(f"Use decorated payment: {decorated_payment}, original: {primaries[0]}")
                        primaries[0] = decorated_payment

```

On main append this `Payment` before we extended `primaries` with `primaries_input`, so there its at index 0. You sure that doesn't have some side effects because now in the continued use of `primaries`, which we iterate over and build the asserts and conditions in order for example, the order of payments is different, i think? 

Maybe it would be good to preserve the old code like in the other comment above, and just replace the payment here if the decorated payment differs from the original one? So, something like the suggestion above. Just not totally sure about the `if len(primaries) > 0 and primaries_input is None:` condition right now and its late.. just wanted to get this out still.